tests/unit/Core/Framework/App/Api/ShopIdControllerTest.php

```diff
@@ -0,0 +1,165 @@
+use Shopware\Core\Framework\App\ShopIdChangeResolver\Resolver;
+use Shopware\Core\Framework\Context;
+use Shopware\Core\Test\Stub\DataAbstractionLayer\StaticEntityRepository;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * @internal
+ */
+#[CoversClass(ShopIdController::class)]
+class ShopIdControllerTest extends TestCase
+{
+    private ShopIdController $controller;
+
+    private Resolver&MockObject $shopIdChangeResolver;
+
+    private ShopIdProvider&MockObject $shopIdProvider;
+
+    /**
+     * @var StaticEntityRepository<AppCollection>
+     */
+    private StaticEntityRepository $appRepository;
+
+    protected function setUp(): void
+    {
+        $this->shopIdChangeResolver = $this->createMock(Resolver::class);
+        $this->shopIdProvider = $this->createMock(ShopIdProvider::class);
+        $this->appRepository = new StaticEntityRepository([]);
+        $this->controller = new ShopIdController($this->shopIdChangeResolver, $this->shopIdProvider, $this->appRepository);
+    }
+
+    public function testGetAvailableStrategies(): void
+    {
+        $this->shopIdChangeResolver->expects($this->once())
+            ->method('getAvailableStrategies')
+            ->willReturn($expectedStrategies = [
+                ['strategy1', 'description1'],
+                ['strategy2', 'description2'],
+            ]);
+
+        $response = $this->controller->getAvailableStrategies();
+        static::assertSame(Response::HTTP_OK, $response->getStatusCode());
+
+        $body = $response->getContent();
+        static::assertIsString($body);
+
+        $strategies = json_decode($body, true, flags: \JSON_THROW_ON_ERROR);
+        static::assertEquals($strategies, $expectedStrategies);
+    }
+
+    public function testChangesShopIdUsingTheProvidedStrategy(): void
+    {
+        $context = Context::createDefaultContext(new AdminApiSource(null));
+        $request = new Request(request: ['strategy' => 'testStrategy']);
+
+        $this->shopIdChangeResolver->expects($this->once())
+            ->method('resolve')
+            ->with($request->request->get('strategy'), $context);
+
+        $response = $this->controller->changeShopId($request, $context);
+        static::assertSame(Response::HTTP_NO_CONTENT, $response->getStatusCode());
+        static::assertEmpty($response->getContent());
+    }
+
+    public function testFailsIfResolverThrowsWhenChangingShopId(): void
+    {
+        $context = Context::createDefaultContext(new AdminApiSource(null));
+        $request = new Request(request: ['strategy' => 'testStrategy']);
+
+        $this->shopIdChangeResolver->expects($this->once())
+            ->method('resolve')
+            ->with($request->request->get('strategy'), $context)
+            ->willThrowException(AppException::shopIdChangeResolveStrategyNotFound('testStrategy'));
+
+        static::expectException(ShopIdChangeStrategyNotFoundException::class);
```

Comment:
here and below. we are trying to use `expectExceptionObject` instead, as it checks everything related